### PR TITLE
chore(ui): migrate AppShellProvider and CodeBlock to TypeScript

### PR DESCRIPTION
**Pull request by gjaskiewicz**
# Summary

Migrate AppShellProvider and CodeBlock to TypeScript

# Changes Made

- migrate AppShellProvider to TypeScript
- migrate CodeBlock to TypeScript
- expose deprecated_js version of AppShellProvider in index.js - temporary fix until this component is removed
- add replacement for CodeBlock in stories (to be replaced after migration of this component)

# Related Issues

- Issue 1: [[link to issue](https://github.com/cloudoperators/juno/issues/212)]
- Issue 2: [[link to issue](https://github.com/cloudoperators/juno/issues/219)]

# Testing Instructions

1. `npm i`
2. `npm run storybook`

# Checklist

- [ ] I have performed a self-review of my code.
- [ ] I have commented my code, particularly in hard-to-understand areas.
- [ ] I have added tests that prove my fix is effective or that my feature works.
- [ ] New and existing unit tests pass locally with my changes.
- [ ] I have made corresponding changes to the documentation (if applicable).
- [ ] My changes generate no new warnings or errors.

### CODE REVIEW

File: packages/ui-components/src/index.js
Comment (andypf) [inline packages/ui-components/src/index.js:10]: double slash

File: packages/ui-components/src/components/AppShellProvider/AppShellProvider.component.tsx
Comment (andypf) [inline packages/ui-components/src/components/AppShellProvider/AppShellProvider.component.tsx:22]: We now have the opportunity to refactor it properly by extracting it into a regular component, outside of this one, without the need for useCallback.

File: packages/ui-components/src/components/AppShellProvider/AppShellProvider.stories.tsx
Comment (andypf) [inline packages/ui-components/src/components/AppShellProvider/AppShellProvider.stories.tsx:7]: can we remove this line?

**Review by barsukov — CHANGES_REQUESTED** [review]
Please fix the Andreas comments and also from my side change index.js to use a deprecated one. 

When we migrate Combobox and also Select we will delete all deprecated related files and change the exports in the index.js

**Comment by barsukov** [comment]
We close caause we need first to migrate the jsonviewer component and also codeblock

File: packages/ui-components/src/components/CodeBlock/CodeBlock.component.tsx
Comment (barsukov) [inline packages/ui-components/src/components/CodeBlock/CodeBlock.component.tsx:157]: Could we get rid of this casting here?

File: packages/ui-components/src/components/CodeBlock/CodeBlock.component.tsx
Comment (barsukov) [inline packages/ui-components/src/components/CodeBlock/CodeBlock.component.tsx:161]: Could we get rid of this casting here?

File: packages/ui-components/src/components/CodeBlock/CodeBlock.component.tsx
Comment (barsukov) [inline packages/ui-components/src/components/CodeBlock/CodeBlock.component.tsx:184]: do we need it here as an object?

**Review by barsukov — CHANGES_REQUESTED** [review]
few comments

File: packages/ui-components/src/components/CodeBlock/CodeBlock.component.tsx
Comment (gjaskiewicz) [reply packages/ui-components/src/components/CodeBlock/CodeBlock.component.tsx]: Done

**Review by barsukov — APPROVED** [review]
LGTM

**Review by TilmanHaupt — APPROVED** [review]

File: packages/ui-components/src/components/CodeBlock/CodeBlock.component.tsx
Comment (guoda-puidokaite) [reply packages/ui-components/src/components/CodeBlock/CodeBlock.component.tsx:133]: Perfect. Don't worry about it in this PR, @gjaskiewicz. Adding this to an improvement EPIC instead.

Comment (guoda-puidokaite) [reply packages/ui-components/src/components/CodeBlock/CodeBlock.component.tsx:133]: Created placeholder issue: https://github.com/cloudoperators/juno/issues/572

File: packages/ui-components/src/components/CodeBlock/CodeBlock.test.tsx
Comment (guoda-puidokaite) [inline packages/ui-components/src/components/CodeBlock/CodeBlock.test.tsx:10]: Would be great during these refactors to improve testing. Some missing edge cases could be:
- Component rendering when both `content` and `children` are `null` or `undefined`.
- Rendering (fallback and default) behaviour for different `lang` prop values, including unsupported languages.
- Clipboard copy functionality to ensure correct text is copied and the "Copied!" message appears.
- Ensuring `timeoutRef` is properly cleared when the component unmounts to prevent memory leaks.
- Ensuring `theCode.current` is correctly assigned and not accessed when `null`.

**Review by guoda-puidokaite — COMMENTED** [review]
Would be great to improve the tests of both components. Especially `AppShellProvider`, it only has a few. If you've the time, of course. 🙈

File: packages/ui-components/src/components/CodeBlock/CodeBlock.component.tsx
Comment (guoda-puidokaite) [reply packages/ui-components/src/components/CodeBlock/CodeBlock.component.tsx:8]: https://github.com/cloudoperators/juno/pull/571#issuecomment-2443769757

**Review by guoda-puidokaite — CHANGES_REQUESTED** [review]
Thank You for the changes!!! 🚀 Last suggested improvements.

File: packages/ui-components/src/components/AppShellProvider/AppShellProvider.component.tsx
Comment (guoda-puidokaite) [inline packages/ui-components/src/components/AppShellProvider/AppShellProvider.component.tsx:18]: I was wondering here if `children` need to be wrapped in a valid JSX tag https://github.com/cloudoperators/juno/pull/547#discussion_r1820773192

**Review by guoda-puidokaite — APPROVED** [review]
Looks good. @barsukov, please double check that recent suggested changes won't cause any issues in your eyes too. 💪

**Review by barsukov — APPROVED** [review]
LGTM when @edda is also ok then we could merge it I guess

**Comment by gjaskiewicz** [comment]
@edda Can you take a look?

**Review by guoda-puidokaite — CHANGES_REQUESTED** [review]
**(Requested Change)**

The `AppShellProvider` has a few breaking changes 🐞:

See screenshots (JS code on left, TS on right)

Screenshot 2024-10-31 at 11 11 45
Screenshot 2024-10-31 at 11 12 07

- Setting `theme-dark` and `theme-light` is broken.  
- `theme` Story **control** has changed from `string` to `object`.
- `theme` **type** has changed from `string` to `string|null`. If the `theme` can only be `theme-light` and `theme-dark`, let's create a **custom type** containing these (instead of just `string`)?
- `theme` **default value** should be `theme-dark`?

**Comment by gjaskiewicz** [comment]
> **(Requested Change)**
> 
> The `AppShellProvider` has a few breaking changes 🐞:
> 
> See screenshots (JS code on left, TS on right)
> 
> Screenshot 2024-10-31 at 11 11 45 Screenshot 2024-10-31 at 11 12 07
> * Setting `theme-dark` and `theme-light` is broken.
> * `theme` Story **control** has changed from `string` to `object`.
> * `theme` **type** has changed from `string` to `string|null`. If the `theme` can only be `theme-light` and `theme-dark`, let's create a **custom type** containing these (instead of just `string`)?
> * `theme` **default value** should be `theme-dark`?

It seems like a potential breaking change in apps to me. If type discrepency is problem in storybook I could just change a control type there. Otherwise it works when theme-light or theme-dark is entered with quotes.

**Comment by guoda-puidokaite** [comment]
@gjaskiewicz I think that the storybook documentation should stay unchanged after the migration. At least to me, it's confusing why I can't just enter text and would have to use quotes, since type is only `string|null`. Please also consider the other suggestions regarding custom type and type default, I think it would be an appropriate time for the improvements.

File: packages/ui-components/src/components/AppShellProvider/AppShellProvider.component.tsx
Comment (edda) [inline packages/ui-components/src/components/AppShellProvider/AppShellProvider.component.tsx:29]:
```suggestion
  theme = "theme-dark",
```

File: packages/ui-components/src/components/AppShellProvider/AppShellProvider.component.tsx
Comment (edda) [inline packages/ui-components/src/components/AppShellProvider/AppShellProvider.component.tsx:59]: We should make explicit in the type what the allowed values are, so rather than `string | null` let's to `"theme-light" | "theme-dark"` and then instead of defaulting to `null`, default to `"theme-dark"`. 
```suggestion
  theme?: "theme-dark" | "theme-light"
```

Comment (edda) [reply packages/ui-components/src/components/AppShellProvider/AppShellProvider.component.tsx:59]: Additionally this type as is breaks the control in storybook (as reported by @guoda-puidokaite )

File: packages/ui-components/src/components/CodeBlock/CodeBlock.component.tsx
Comment (edda) [reply packages/ui-components/src/components/CodeBlock/CodeBlock.component.tsx:121]: Please fix as suggested by @guoda-puidokaite

File: packages/ui-components/src/components/StyleProvider/StyleProvider.component.tsx
Comment (edda) [inline packages/ui-components/src/components/StyleProvider/StyleProvider.component.tsx:152]: See comment above. Change to
```suggestion
  theme?: "theme-dark" | "theme-light"
```
Though ideally, since we use this in two places we define the type once and use it in both places.

**Review by edda — DISMISSED** [review]

**Comment by gjaskiewicz** [comment]
> @gjaskiewicz Nice that it's just a Storybook issue. I think that the storybook documentation should stay unchanged though after the migration. At least to me, it's confusing why I can't just enter text and would have to use quotes, since `type` is only `string|null`. Please also consider the other suggestions regarding **custom type** and **type default**, I think it would be an appropriate time for the improvements. 🚀

We could go with 3 options here:
- set it to enumerated type, if Esther confirms that setting arbitrary string should be no longer supported,
- allow arbitrary string, but disallow null, we could set default value to "theme-dark" then
- leave type as is, but change control in strorybook to string control rather than object control

As some of the options will cause change of behavior of the component it would be good if also @barsukov and @edda express their opinion

File: packages/ui-components/src/components/StyleProvider/StyleProvider.component.tsx
Comment (guoda-puidokaite) [reply packages/ui-components/src/components/StyleProvider/StyleProvider.component.tsx:152]: Indeed, I'd suggest a **custom type**, as [above](https://github.com/cloudoperators/juno/pull/547#issuecomment-2449722013) .

Comment (gjaskiewicz) [reply packages/ui-components/src/components/StyleProvider/StyleProvider.component.tsx:152]: As discussed f2f - I changed the type

File: packages/ui-components/src/components/CodeBlock/CodeBlock.component.tsx
Comment (guoda-puidokaite) [reply packages/ui-components/src/components/CodeBlock/CodeBlock.component.tsx:188]: This was mentioned as a future improvement. Added [here](https://github.com/cloudoperators/juno/issues/570).

**Comment by guoda-puidokaite** [comment]
(**Improvement**)

For `AppShellProvider`, the `children` type is an `array` in `Storybook` but `React.ReactNode` in the code.

I think we could improve this for `AppShellProvider` (left old .JS code, right new .TSX code) :
Screenshot 2024-11-04 at 15 13 27

In other components, it's `any`, as Storybook doesn't have an appropriate type with a comment. 
Example of what we do in `Message`:

**Review by guoda-puidokaite — APPROVED** [review]
Small comment on types in Storybook, but idk if necessary to change.
Otherwise Thank You for the type changes! Looks great. 🚀💪🤪

File: packages/ui-components/src/components/AppShellProvider/AppShellProvider.component.tsx
Comment (barsukov) [inline packages/ui-components/src/components/AppShellProvider/AppShellProvider.component.tsx:44]: little comment from my side you can use it as absed props for AppShellProviderProps and extends it removing them from AppShellProviderProps

**Review by barsukov — COMMENTED** [review]
I left one comment that kinda nice to fix before we merge.

File: packages/ui-components/src/components/AppShellProvider/AppShellProvider.component.tsx
Comment (gjaskiewicz) [reply packages/ui-components/src/components/AppShellProvider/AppShellProvider.component.tsx:44]: Done

**Review by guoda-puidokaite — APPROVED** [review]
LGTM too! 🚀 Thank You!

**Pull request merged**
